### PR TITLE
allow external control of gradient accumulation boundary

### PR DESCRIPTION
**Pull request by jeffra**
Many integrations are made a lot easier by deferring to the client code for when we are in a gradient accumulation boundary or not. This PR allows the client code to manually set the boundary state and propagates it to the right places in the DS engine.

`self.optimizer` may optionally use the `is_gradient_accumulation_boundary` state, but no harm in passing the state there for non-zero optimizers as well.

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -1746,11 +1747,18 @@ def is_gradient_accumulation_boundary(self):
...
             bool: if the current step is a gradient accumulation boundary.
         """
-        return (self.micro_steps + 1) % self.gradient_accumulation_steps() == 0
+        if self._is_gradient_accumulation_boundary is None:
+            return (self.micro_steps + 1) % \
+                self.gradient_accumulation_steps() == 0
+        else:
+            return self._is_gradient_accumulation_boundary
+
+    def set_gradient_accumulation_boundary(self, is_boundary):
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:1759]: I think a doc string is needed here to help clients use properly. In particular, about ensuring that clients clear this flag after optimizer.step().

Comment (tjruwase) [reply deepspeed/runtime/engine.py:1759]: An alternative approach that comes to mind is instead of adding a new engine function, we can add a `is_gradient_accumulation_boundary` argument to `backward()` and `step()` such as: 
```python
   def backward(self, ..., is_gradient_accumulation_boundary=False):
         self._is_gradient_accumulation_boundary = is_gradient_accumulation_boundary
        ...

  def step(self, ..., is_gradient_accumulation_boundary=False):
        self._is_gradient_accumulation_boundary = is_gradient_accumulation_boundary
       ...
```
I am not sure if this will work well with the current kwargs constraints of `backward` and `step`.

File: deepspeed/runtime/engine.py [hunk 2]
```diff
@@ -1757,6 +1757,26 @@ def is_gradient_accumulation_boundary(self):
...
+
+        engine.set_gradient_accumulation_boundary(False)
+        for _ in range(gradient_accumulation_steps - 1):
+            micro_batch = next(data_loader)
+            loss = engine(micro_batch)
+            engine.backward(loss)
+        engine.set_gradient_accumulation_boundary(True)
+        micro_batch = next(data_loader)
+        loss = engine(micro_batch)
+        engine.backward(loss)
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:1775]: For completeness, should this snippet not include `engine.step()`?

Comment (jeffra) [reply deepspeed/runtime/engine.py:1775]: Oh good call, that's an important bit, in this case we only need to call `engine.step()` once at the end.

Comment (tjruwase) [reply deepspeed/runtime/engine.py:1776]: Also, would the following be correct usage of this api?
```python
   for i, micro_batch in enumerate(next(data_loader)):
        is_gradient_accumulation_boundary = (i+1 % gradient_accumulation_steps) == 0
        engine.set_gradient_accumulation_boundary(is_gradient_accumumation_boundary)
        loss = engine(micro_batch)
        engine.backward(loss)
        engine.step()

```

Comment (jeffra) [reply deepspeed/runtime/engine.py:1776]: Yep, that would work too. In the case of megatron and fairseq they both call their optimizer step in completely different code paths then the main fwd/bwd passes so it's a bit more aligned with the style in the docstring (without a final engine.step) I guess.

**Review by tjruwase — APPROVED** [review]
